### DISCUSSION TITLE
Meta Scripts in multiple Shells? Or suspend via remote control?

### DISCUSSION THREAD

**Discussion by hainkind** [kovidgoyal/kitty#6796] [Q&A] [answered]
Hi,
i'm probably not using the right words in the title so let me explain. I'm currently learning Pentesting and I often find myself doing a sequence of commands in multiple shells within the same terminal and i'm wondering if i can automate this somehow. 

A simple example would be upgrading a nc reverse shell:

```
# In nc reverse shell
$ python -c 'import pty; pty.spawn("/bin/bash")' # upgrade to tty
Ctrl-Z suspend

# In Kali
$ stty -a # get term size
$ stty raw -echo; fg

# In reverse shell
$ reset
$ export SHELL=bash
$ export TERM=xterm-256color
$ stty rows <num> columns <cols>
```

i already looked at the documentation and it seems like it might be possible with remote control but i haven't found anything about triggering suspend. i'd be grateful for any hints.

**Comment by kovidgoyal** [comment]
https://sw.kovidgoyal.net/kitty/kittens/broadcast/

**Comment by hainkind** [comment]
thanks for the pointer.
but i'm wondering how to script the ctrl+z suspend. when i'm using the kitten, the suspend does get broadcasted but is there a simple way to script this? I tried ^Z but i get broadcasted literally

or maybe i wasn't clear. i don't need to do this in multiple shells in parallel, i rather want to map this sequence to a key

**Comment by kovidgoyal** [comment]
Works fine for me, the ctrl-z triggers a SIGSUSP if the
destination window has not turned of ISIG in termios, as expected.
Run cat in your destination window and watch it get suspended.

**Comment by kovidgoyal** [comment]
Oh and if you want to send a signal to a window, without relying on termios state, use kitty @ signal-child

**Reply by hainkind** [reply]
ok i think i got it.
 i just found out about the show_key kitten
```
kali@kali:~$ kitty +kitten show_key
Press any keys - Ctrl+D will terminate this program
UNIX		send_text
^Z		\x1a 
```

so something like
```
kitty @ send-text python -c 'import pty; pty.spawn("/bin/bash")\r

kitty @ signal-child SIGTSTP
# or
kitty @ send-text \x1a

kitty @ send-text stty -a\r
# parse term size

```
etc. should do the trick?
i will try to script it tonight.

oh and while trying i found this, seems like SIGSUSP is not implemented?

```
kali@kali:~$ kitty @ signal-child SIGSUSP         
Traceback (most recent call last):
  File "kitty/boss.py", line 704, in _execute_remote_command
  File "kitty/remote_control.py", line 220, in handle_cmd
  File "kitty/rc/signal_child.py", line 43, in response_from_kitty
  File "kitty/rc/signal_child.py", line 43, in <genexpr>
AttributeError: module 'signal' has no attribute 'SIGSUSP'

Error: module 'signal' has no attribute 'SIGSUSP'
```

thanks for your help and your amazing work here.

**Comment by kovidgoyal** [comment]
typo SIGSUSP = SIGSTOP

**Comment by hainkind** [comment]
```
#kitty.conf
map alt+u launch --type=overlay --allow-remote-control ~/upgrade_nc_shell.sh @active-kitty-window-id

#upgrade_nc_shell.sh
#!/usr/bin/env bash
window_id=$1
kitty @ send-text --match id:$window_id "python3 -c 'import pty; pty.spawn(\"/bin/bash\")'\n"
kitty @ signal-child --match id:$window_id SIGTSTP 
kitty @ send-text --match id:$window_id "stty -a\n" 
output=$(kitty @ get-text --match id:$window_id)
rows=$(echo "$output" | grep -oP 'rows \K\d+')
columns=$(echo "$output" | grep -oP 'columns \K\d+')
kitty @ send-text --match id:$window_id "stty raw -echo; fg\n"
kitty @ send-text --match id:$window_id "export SHELL=bash\n"
kitty @ send-text --match id:$window_id "export TERM=xterm-256color\n"
kitty @ send-text --match id:$window_id "stty rows $rows columns $columns\n"
kitty @ send-text --match id:$window_id "reset\n"
```
I managed to get it to work like this but to me it seems that launch type=background would be more appropriate but i can't make it work. i don't understand how to target the window/tab i'm launching from with the background script.

> background
The process will be run in the background, without a kitty window. Note that if kitten @ launch --allow-remote-control is specified the KITTY_LISTEN_ON environment variable will be set to a dedicated socket pair file descriptor that the process can use for remote control.

i don't get how to use $KITTY_LISTEN_ON in this context. do you have a tip for me?

**Comment by kovidgoyal** [comment, marked as answer]
KITTY_LISTEN_ON is automatically used by kitten @

And you run

kitten @ ls --match state:active

to get the currently active window. Also use the remote_script action
added in kitty 0.31.0

**Reply by hainkind** [reply]
ok amazing in 0.31.0 it works with type=background and --match state:active!
thanks alot!
